### PR TITLE
Eamon/line items

### PR DESCRIPTION
**Pull request by eito-fis**
Adds grouping models to `encoding/target_encoders.py`, `target_models/grouping.py`, `nn/group_target_blocks.py`, `tests/test_group.py` and `tests/test_group_target_encoder.py`. Also updates `fused_featurizer()` to apply to non-T5 base models, and adds the `FusedRoBERTa` and `FusedDocRep` base models. Finally, updates the huggingface base models to use an updated version of transformers (v4.5.1) that is compatible with the new fast tokenizer changes.

### CODE REVIEW

File: finetune/util/featurizer_fusion.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        subsize = config.max_length // config.num_fusion_shards
+        context = kwargs.pop("context")
+        for start in range(0, config.max_length, subsize):
+            xi = X[:, start : start + subsize]
+            lengths = tf.maximum(lengths - tf.shape(xi)[1], 0)
+            if context is not None:
+                ci = context[:, start : start + subsize]
+            else:
+                ci = None
+            # if config.low_memory_mode and train:
```
Comment (madisonmay) [inline finetune/util/featurizer_fusion.py]: Should this be removed or uncommented?

File: finetune/base_models/bert/modeling.py [hunk 2]
```diff
@@ -246,8 +246,14 @@ def __init__(
             with tf.compat.v1.variable_scope("pooler"):
                 # We "pool" the model by simply taking the hidden state corresponding
                 # to the first token. We assume that this has been pre-trained
-
-                first_token_tensor = tf.squeeze(self.sequence_output[:, 0:1, :], axis=1)
+                output_shape = tf.shape(self.sequence_output)
```
Comment (madisonmay) [inline finetune/base_models/bert/modeling.py:249]: This is done because there's some small chance we generate a chunk w/ no valid tokens?

File: finetune/target_models/seq2seq.py [hunk 3]
```diff
@@ -1,11 +1,57 @@
...
 from finetune.encoding.target_encoders import Seq2SeqLabelEncoder
 from finetune.input_pipeline import BasePipeline
 from finetune.model import PredictMode
-from finetune.util.beam_search import beam_search
+from finetune.util.beam_search import beam_search, get_state_shape_invariants
+from tensorflow.python.util import nest
+import itertools
+import sys
+
+def py_bracket_constraint(history, bracket_idx_pairs, eos_token, vocab_size):
```
Comment (benleetownsend) [inline finetune/target_models/seq2seq.py]: I think this stuff can be removed.

File: finetune/base_models/bert/featurizer.py [hunk 4]
```diff
@@ -49,7 +52,7 @@ def bert_featurizer(
         hidden_act=config.act_fn,
         hidden_dropout_prob=config.resid_p_drop,
         attention_probs_dropout_prob=config.attn_p_drop,
-        max_position_embeddings=config.max_length, 
+        max_position_embeddings=max_length,
```
Comment (madisonmay) [inline finetune/base_models/bert/featurizer.py:55]: Wild -- there are BERT position embeddings up to 2048 (even for small model architecture)?! Curious if they are trained or just randomly initialized / never used during pre-training.

File: finetune/encoding/target_encoders.py [hunk 5]
```diff
@@ -135,6 +136,45 @@ def transform(self, y):
     def inverse_transform(self, y):
         return [self.encoder.decode(y_i.tolist()) for y_i in y]
 
+class SequenceLabelingTextEncoder(Seq2SeqLabelEncoder):
+    def transform(self, y):
+        all_labels = []
+        for y_i in y:
+            # List of dictionaries, where the key is the label and the value is
+            # the span being labeled
+            labels = [{span["label"]: span["text"]} for span in y_i]
```
Comment (madisonmay) [inline finetune/encoding/target_encoders.py]: Maybe note that there's an expectation these follow some sort of order for future interns to avoid potential sadness.

File: finetune/encoding/target_encoders.py [hunk 6]
```diff
@@ -239,21 +300,461 @@ def transform(self, out, labels):
...
+            group_labels.append({
+                "start": group_start,
+                "end": group_end,
+                "label": "",
+                "text": group_text,
+                "group_start": True,
+            })
+
+        # Encode grouping labels
+        self.classes_ = self.group_classes_
```
Comment (madisonmay) [inline finetune/encoding/target_encoders.py]: The set state on self --> call super() pattern is a little bit gross, part of me thinks this would be cleaner if it were more functional  + settings like bio_tagging / group_tagging could be passed to .transform().

File: finetune/encoding/target_encoders.py [hunk 7]
```diff
@@ -239,21 +300,461 @@ def transform(self, out, labels):
...
+                        start_token_labels[i] = self.lookup[current_tag]
+                        prev_idx = i
+                    elif prev_idx is not None:
+                        next_token_labels[prev_idx] = i
+                        prev_idx = i
+
+        return [start_token_labels, next_token_labels]
+
+    def inverse_transform(self, y):
+        # TODO: Fix this breaking class weights
```
Comment (madisonmay) [inline finetune/encoding/target_encoders.py]: Is this TODO still relevant?

File: finetune/util/featurizer_fusion.py (hunk 1 above)
Comment (benleetownsend) [reply finetune/util/featurizer_fusion.py]: we should keep this in for doc2dict but probably make it configurable.

File: finetune/encoding/input_encoder.py [hunk 8]
```diff
@@ -177,6 +177,7 @@ def _cut_and_concat(
         delimiter=None,
         end=None,
         include_bos_eos=True,
+        eos_on_cut=True,
```
Comment (benleetownsend) [inline finetune/encoding/input_encoder.py:175]: Can probably cut this - I think this is now unused?

File: finetune/encoding/target_encoders.py [hunk 9]
```diff
@@ -135,6 +136,45 @@ def transform(self, y):
     def inverse_transform(self, y):
         return [self.encoder.decode(y_i.tolist()) for y_i in y]
 
+class SequenceLabelingTextEncoder(Seq2SeqLabelEncoder):
```
Comment (benleetownsend) [inline finetune/encoding/target_encoders.py]: Doesn't have to be now but we should split this file up sometime

File: finetune/encoding/target_encoders.py [hunk 10]
```diff
@@ -135,6 +136,45 @@ def transform(self, y):
...
+    def transform(self, y):
+        all_labels = []
+        for y_i in y:
+            # List of dictionaries, where the key is the label and the value is
+            # the span being labeled
+            labels = [{span["label"]: span["text"]} for span in y_i]
+            all_labels.append(json.dumps(labels))
+        ret = super().transform(all_labels)
+        return ret
```
Comment (benleetownsend) [inline finetune/encoding/target_encoders.py]: This takes a dict as input but returns the strings as predictions? Could we maybe add a TODO here to add the inverse_transform. (or add it?)

File: finetune/util/group_metrics.py [hunk 11]
```diff
@@ -0,0 +1,522 @@
+from copy import deepcopy
+from collections import namedtuple
+
+import numpy as np
+from scipy.optimize import linear_sum_assignment
```
Comment (madisonmay) [inline finetune/util/group_metrics.py:5]: Very neat!  Wasn't familiar with this.

File: finetune/util/group_metrics.py [hunk 12]
```diff
@@ -0,0 +1,522 @@
...
+    counts = [[(0, 0, 0) for _ in len(labels)] for _ in len(preds)]
+    costs = [[0 for _ in len(labels)] for _ in len(preds)]
+
+    for i, pred in enumerate(preds):
+        for j, label in enumerate(labels):
+            TP, FP, FN = count_fn(pred, label)
+            counts[i][j] = (TP, FP, FN)
+            costs[i][j] =  -(TP - (FP + FN) / 2)
+
+    pred_idxs, label_idxs = linear_sum_assignment(costs)
```
Comment (madisonmay) [inline finetune/util/group_metrics.py:251]: I don't think this quite works.  If `costs.shape[0] != costs.shape[1]`, I don't think the counts won't be the desired TP / FP / FN counts (will underestimate FP/FN?).  You can maybe force the costs + counts matrices to be square and fill the rest of the matrix with FP (in case of `len(preds) > len(labels)`) or FN (in case of `len(preds) < len(labels)`)?  Personally also think it makes sense to first group by label before running these calcs?

Comment (madisonmay) [reply finetune/util/group_metrics.py:251]: Take this with a grain of salt because I need to stare at longer

File: finetune/util/group_metrics.py [hunk 13]
```diff
@@ -0,0 +1,522 @@
...
+                label_label = doc_labels[label_idx]["label"]
+
+                class_counts[label_label]["true_positives"] += TP
+                # False positives attributed to the predicted label
+                class_counts[pred_label]["false_positives"] += FP
+                # False negatives attributed to the ground truth label
+                class_counts[label_label]["false_negatives"] += FN
+
+        pred_idxs, label_idxs = list(zip(*assign_idxs))
+        if len(doc_preds) > len(doc_labels):
```
Comment (madisonmay) [inline finetune/util/group_metrics.py:310]: Oh you account for this later.  Nice.

Comment (madisonmay) [reply finetune/util/group_metrics.py:310]: I do think this would be much cleaner if it was done on a per class basis, but I *think* this works out fine just tricky.

File: finetune/util/group_metrics.py [hunk 14]
```diff
@@ -0,0 +1,522 @@
...
+    else:
+        return per_class_metrics
+
+def entity_in_group(entity, group):
+    """
+    Returns True if entity is contained in a group span, False otherwise
+    :param entity: A single entity span
+    :param group: A single group
+    """
+    for span in group["tokens"]:
```
Comment (madisonmay) [inline finetune/util/group_metrics.py]: Use of "token" to refer to a group span is kind of confusing, makes me think these are BPEs tokens or something.

File: finetune/util/group_metrics.py [hunk 15]
```diff
@@ -0,0 +1,522 @@
...
+        if len(doc_preds) > len(doc_labels):
+            missing_idxs = set(range(len(doc_preds))) - set(pred_idxs)
+            extra_groups = doc_preds
+            # All examples in extra pred groups are false positives
+            add_count = "false_positives"
+        elif len(label_groups) > len(pred_groups):
+            missing_idxs = set(range(len(doc_labels))) - set(label_idxs)
+            extra_groups = doc_labels
+            # All examples in extra label groups are false negatives
+            add_count = "false_negatives"
```
Comment (madisonmay) [inline finetune/util/group_metrics.py:319]: I'm not sure this accounts for the scenario where one class has a surplus of predictions and another has a deficit?

File: finetune/util/group_metrics.py (hunk 14 above)
Comment (eito-fis) [reply finetune/util/group_metrics.py]: Yeah this is something that happened way early in the project and totally doesn't match what it is. Hopefully can find time to fix this today or tomorrow

File: finetune/util/group_metrics.py (hunk 15 above)
Comment (eito-fis) [reply finetune/util/group_metrics.py:319]: Hmm, in that case I think it would be handled during the earlier group assignment step? We would do our best to match groups that have the same label but would probably be forced to make some assignments that aren't optimal. Regardless, I think it holds that all un-matched predictions are false positives and all un-matched ground truth labels are false negatives.

File: finetune/util/group_metrics.py (hunk 13 above)
Comment (eito-fis) [reply finetune/util/group_metrics.py:310]: Per class might be nicer - I do think one benefit from how it's written right now is that we can switch to punish group level label mismatches less if we wanted. Ex: treating a label mismatch as an addition to the FP and FN counts instead of fully throwing out the groups.

Comment (madisonmay) [reply finetune/util/group_metrics.py:310]: Makes sense, thanks for the comment!

File: finetune/util/group_metrics.py (hunk 15 above)
Comment (madisonmay) [reply finetune/util/group_metrics.py:319]: Yup that checks out -- sorry for the false alarm, just took me a while to wrap my head around this one.

File: finetune/base_models/bert/modeling.py (hunk 2 above)
Comment (eito-fis) [reply finetune/base_models/bert/modeling.py:249]: Yeah, with featurizer fusion theres chance that we pass 0 length chunks to the featurizer. Theres a chance we can avoid this with by checking for the invalid chunks in the featurizer fusion function, but I wasn't able to get it to work before. Will give it another shot.

Comment (eito-fis) [reply finetune/base_models/bert/modeling.py:249]: Unfortunately a bit more complicated then expected, as there are a lot of values returned by the featurizer that vary from base model to base model, and we have to replicate them all if we want to write a conditional in featurizer_fusion(). Won't be able to get it done today, but someone might want to take a look at a later point.

File: finetune/base_models/bert/featurizer.py (hunk 4 above)
Comment (eito-fis) [reply finetune/base_models/bert/featurizer.py:55]: Not sure haha, didn't think about this

File: finetune/encoding/target_encoders.py (hunk 6 above)
Comment (eito-fis) [reply finetune/encoding/target_encoders.py]: Totally agree that its gross, but was mainly trying to avoid adding more code to the Sequence Labeling Encoder. Would have liked to come up with a better solution though

File: finetune/encoding/target_encoders.py (hunk 10 above)
Comment (eito-fis) [reply finetune/encoding/target_encoders.py]: Good call, had the logic in the target model but putting it as an inverse transform makes much more sense.
